std/range/package.d

```diff
@@ -4173,10 +4195,10 @@ private string lockstepMixin(Ranges...)(bool withIndex)
             }
             return res;
         }
-    }, params.join(", "), withIndex ? "size_t index = 0;" : "",
+    }, name, params.join(", "), indexDef,
        emptyChecks.join(" && "), dgArgs.join(", "),
        popFronts.join("\n                "),
-       withIndex ? "index++;" : "");
```

Comment:
I guess demanding allSatisfy!(isBidirectionalRange, Ranges) && allSatisfy!(hasLength, Ranges) to use the indexed version should be good enough. And add in an assert like in the forward-iterating version.